File: 100-Days-Series/Day 75/RemoveDuplicatesfromSortedList.js
class LinkedList {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class RemoveDuplicatesFromSortedList {
    constructor() {
        this.head = null; // Head of the linked list
    }

    // Method to add a new node
    addNode(data) {
        const newNode = new LinkedList(data);
        if (this.head === null) {
            this.head = newNode;
        } else {
            let temp = this.head;
            while (temp.next !== null) {
                temp = temp.next;
            }
            temp.next = newNode;
        }
    }

    // Method to traverse the linked list
    traverse() {
        if (this.head === null) {
            console.log("Linked List is Empty");
        } else {
            let temp = this.head;
            while (temp !== null) {
                console.log("Data is " + temp.data);
                console.log("------------------------------");
                temp = temp.next;
            }
        }
    }

    // Method to remove duplicates
    removeDuplicates() {
        if (this.head === null) {
            console.log("Linked List is Empty");
        } else {
            let current = this.head;
            while (current.next !== null) {
                if (current.data === current.next.data) {
                    current.next = current.next.next;
                } else {
                    current = current.next;
                }
            }
        }
    }
}

// Main Functionality
const list = new RemoveDuplicatesFromSortedList();

list.addNode(10)
list.addNode(20)
list.addNode(20)
list.addNode(20)
list.addNode(30)
list.addNode(40)
list.addNode(50)
list.traverse()
console.log("After Removing Duplicate Node");
list.removeDuplicates();
list.traverse();

File: 100-Days-Series/Day 75/RemoveDuplicatesfromSortedList.cpp
#include <iostream>
using namespace std;

class LinkedList {
public:
    int data;
    LinkedList* next;

    // Constructor
    LinkedList(int data) {
        this->data = data;
        this->next = nullptr;
    }
};

class Example {
public:
    LinkedList* head;

    // Constructor to initialize head
    Example() {
        head = nullptr;
    }

    // Add a node to the linked list
    void addNode() {
        int value;
        cout << "Enter data: ";
        cin >> value;

        LinkedList* newNode = new LinkedList(value);
        if (head == nullptr) {
            head = newNode;
        } else {
            LinkedList* temp = head;
            while (temp->next != nullptr) {
                temp = temp->next;
            }
            temp->next = newNode;
        }
    }

    // Traverse the linked list
    void traverse() {
        if (head == nullptr) {
            cout << "Linked List is empty!" << endl;
        } else {
            LinkedList* temp = head;
            while (temp != nullptr) {
                cout << "Data: " << temp->data << endl;
                temp = temp->next;
            }
        }
    }
    // Function to remove duplicates
    void removeDuplicates(){
        if (head == nullptr || head->next == nullptr) {
            return; // Nothing to remove
        }
        LinkedList* current = head;
        while (current->next != nullptr) {
            if (current->data == current->next->data) {
                LinkedList* duplicate = current->next;
                current->next = current->next->next;
                delete duplicate; // Free memory of duplicate node
            } else {
                current = current->next;
            }
        }
    }
};
// Let's Start Here
int main() {
    Example emp;

    int choice;
    do {
        cout << "\n1. Add Node";
        cout << "\n2. Traverse";
        cout << "\n3. Remove Duplicates";
        cout << "\n4. Exit";
        cout << "\nEnter your choice: ";
        cin >> choice;

        switch (choice) {
            case 1:
                emp.addNode();
                break;
            case 2:
                emp.traverse();
                break;
            case 3:
                emp.removeDuplicates();
                break;
            case 4:
                cout << "Exiting..." << endl;
                break;
            default:
                cout << "Invalid choice!" << endl;
        }
    } while (choice != 4);

    return 0;
}
